File: lcof/面试题24. 反转链表/Solution.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // 通过头插实现逆序
        // ListNode *first = new ListNode(-1);
        // ListNode *p = head, *q;
        // while (p) {
        //     q = p->next;
        //     p->next = first->next;
        //     first->next = p;
        //     p = q;
        // }
        // return first->next;

        // 常规方法
        ListNode *pre = NULL, *cur = head;
        while (cur) {
            ListNode* temp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = temp;
        }
        return pre;
    }
};

File: lcof/面试题24. 反转链表/Solution.swift
/* public class ListNode {
*     var val: Int
*     var next: ListNode?
*     init(_ val: Int) {
*         self.val = val
*         self.next = nil
*     }
* }
*/

class Solution {
    func reverseList(_ head: ListNode?) -> ListNode? {
        let dummy = ListNode(0)
        var curr = head
        
        while curr != nil {
            let next = curr?.next
            curr?.next = dummy.next
            dummy.next = curr
            curr = next
        }
        
        return dummy.next
    }
}